support-frontend/assets/helpers/redux/checkout/product/subscriptionsSideEffects.ts

```diff
@@ -0,0 +1,49 @@
+import { isAnyOf } from '@reduxjs/toolkit';
+import {
+	finalPrice,
+	getProductPrice,
+} from 'helpers/productPrice/productPrices';
+import type { SubscriptionsStartListening } from 'helpers/redux/subscriptionsStore';
+import {
+	setBillingPeriod,
+	setDiscountedPrice,
+	setProductOption,
+	setProductPrices,
+	setProductType,
+	setSelectedProductPrice,
+} from './actions';
+
+const mayChangeProductPrice = isAnyOf(
+	setProductType,
+	setBillingPeriod,
+	setProductOption,
+	setProductPrices,
+);
+
+export function addProductSideEffects(
+	startListening: SubscriptionsStartListening,
+): void {
+	startListening({
+		matcher: mayChangeProductPrice,
+		effect(_, listenerApi) {
```

Comment:
We currently calculate and re-calculate the actual relevant `ProductPrice` object, and a secondary version with any relevant discount applied, in multiple places in store-connected components. This listener enables us to keep that information in Redux instead, and automatically update it if the user changes billing period (possible in the GW checkout), product type (possible by 'adding the digital subscription' in the paper checkout) and so on.

Reply:
Yeah I think it makes a lot of sense to bring this all in to the store instead of recalculating many times. 

This might be my lack of familiarity with the new rtk stuff, but it doesn't necessarily seem 'side-effecty'? i.e updating the discounted price should be a direct effect of calling `setProductType`, and therefore could happen in the reducer?

Reply:
I put it in side effects because it may need to change based multiple variables beyond just changing the product type. Also I think it's simpler to stick to both reducers only affecting the state of a single property at a time as far as possible, and having a single way of saying 'when actions a, b or c happen, also do something else'.

Reply:
I guess it makes the reducers themselves simpler, but perhaps makes following the overall logic less simple? If all of the state changes are within the reducer, then it's a bit easier to understand everything that happens when an action is dispatched. It's just a little bit harder to understand the effects if you have to also look for all of the side effects that are waiting to trigger on the dispatch.

Reply:
That's fair, I suppose in this case since the change is within the same reducer slice it's much less 'side-effecty' than having to reach out to another part of state or perform some effect outside of Redux's scope like putting something in localStorage. I'll have a go at reworking this to happen within the reducer and see how it looks.

Reply:
I've just had a go and ran into the immediate problem that we can't access the current `countryId` from inside the product slice- reducers can only see the piece of state they control, and internationalisation info is managed in the `common` slice. Listeners, however, have access to the whole store as it is at the point a given action happens. We can't determine the relevant product price without knowing what country the user is in.